ledger/eval.go

```diff
@@ -1088,6 +1164,56 @@ func (eval *BlockEvaluator) finalValidation() error {
 				return fmt.Errorf("CompactCertType %d unexpected", ccType)
 			}
 		}
+
+		proto, ok := config.Consensus[eval.block.BlockHeader.CurrentProtocol]
+		if !ok {
+			return fmt.Errorf("unknown consensus version: %s", eval.block.BlockHeader.CurrentProtocol)
+		}
+
+		expectedMaxNumberOfExpiredAccounts := proto.MaxExpiredAccountsToProcess
+		lengthOfExpiredParticipationAccounts := len(eval.block.ParticipationUpdates.ExpiredParticipationAccounts)
+
+		// If the length of the array is strictly greater than our max then we have an error.
+		// This works when the expected number of accounts is zero (i.e. it is disabled) as well
+		if lengthOfExpiredParticipationAccounts > expectedMaxNumberOfExpiredAccounts {
+			return fmt.Errorf("length of expired accounts (%d) was greater than expected (%d)",
+				lengthOfExpiredParticipationAccounts, expectedMaxNumberOfExpiredAccounts)
+		}
+
+		// For security reasons, we need to make sure that all addresses in the expired participation accounts
+		// are unique.  We make this map to keep track of previously seen address
+		addressSet := make(map[basics.Address]bool)
```

Comment:
preallocate with size : 
initialize the map to have `eval.block.ParticipationUpdates.ExpiredParticipationAccounts` entries.